File: ruby/test/test_scanner.rb
# frozen_string_literal: true

require "minitest/autorun"
require "tmpdir"
require "fileutils"
require_relative "../lib/envdoctor/scanner"
require_relative "../lib/envdoctor/cli"

class ScannerTest < Minitest::Test
  def test_detects_usage_and_ignores_comments
    src = <<~RUBY
      # ENV["COMMENTED"]
      db = ENV["DB_URL"]
      port = ENV.fetch("PORT")
      =begin
      ENV["BLOCK_IGNORED"]
      =end
      user = ENV['DB_USER']
    RUBY
    used = Envdoctor::Scanner.scan_source("app.rb", src)
    assert_equal %w[DB_URL DB_USER PORT], used.keys.sort
    refute used.key?("COMMENTED")
    refute used.key?("BLOCK_IGNORED")
  end

  def test_reconciles_missing_and_unused
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "DB_URL=x\nUNUSED_KEY=1\n")
      File.write(File.join(dir, "app.rb"), "ENV[\"DB_URL\"]\nENV[\"NEW_FLAG\"]\n")
      findings = Envdoctor::Scanner.scan(dir)
      errors = findings.select { |f| f.severity == "error" }.map(&:name)
      warnings = findings.select { |f| f.severity == "warning" }.map(&:name)
      assert_includes errors, "NEW_FLAG"
      assert_includes warnings, "UNUSED_KEY"
      refute_includes errors, "DB_URL"
      refute_includes warnings, "DB_URL"
    end
  end

  def test_detects_duplicate_keys
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "DUP=a\nSOLO=1\nDUP=b\n")
      File.write(File.join(dir, "app.rb"), "ENV[\"DUP\"]\nENV[\"SOLO\"]\n")
      findings = Envdoctor::Scanner.scan(dir)
      dup = findings.find { |f| f.rule == "duplicates" }
      refute_nil dup
      assert_equal "DUP", dup.name
      assert_equal "error", dup.severity
      assert_equal "defined 2 times in the same file (lines 1, 3)", dup.message
      # single-definition key not reported as duplicate
      refute(findings.any? { |f| f.rule == "duplicates" && f.name == "SOLO" })
      # duplicated-but-used var must NOT also be flagged unused/undefined
      refute(findings.any? { |f| f.name == "DUP" && %w[unused undefined-in-source].include?(f.rule) })
    end
  end

  def test_detects_public_prefix_secrets
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"),
                 "NEXT_PUBLIC_API_KEY=x\nVITE_SECRET=y\nPUBLIC_URL=z\nAPI_KEY=w\nPUBLIC_KEY=k\n")
      findings = Envdoctor::Scanner.scan(dir)
      flagged = findings.select { |f| f.rule == "public-prefix" }.map(&:name)
      assert_includes flagged, "NEXT_PUBLIC_API_KEY"
      assert_includes flagged, "VITE_SECRET"
      refute_includes flagged, "PUBLIC_URL"
      refute_includes flagged, "API_KEY"
      refute_includes flagged, "PUBLIC_KEY"
      flagged.each do |name|
        f = findings.find { |x| x.rule == "public-prefix" && x.name == name }
        assert_equal "error", f.severity
      end
    end
  end

  def test_detects_weak_secret
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"),
                 "API_KEY=changeme\nDB_PASSWORD=s3cureLongValue!\nAUTH_TOKEN=\nSECRET_KEY=xxxxxx\nDATABASE_URL=postgres://localhost/db\n")
      File.write(File.join(dir, "app.rb"),
                 "ENV[\"API_KEY\"]\nENV[\"DB_PASSWORD\"]\nENV[\"AUTH_TOKEN\"]\nENV[\"SECRET_KEY\"]\nENV[\"DATABASE_URL\"]\n")
      findings = Envdoctor::Scanner.scan(dir)
      weak = findings.select { |f| f.rule == "weak-secret" }
      names = weak.map(&:name)
      assert_includes names, "API_KEY"      # placeholder value
      assert_includes names, "AUTH_TOKEN"   # empty value
      assert_includes names, "SECRET_KEY"   # xxxxxx placeholder
      refute_includes names, "DB_PASSWORD"  # strong value
      refute_includes names, "DATABASE_URL" # not a secret-looking name
      weak.each do |f|
        assert_equal "warning", f.severity
        assert_equal "secret-looking variable has a weak or placeholder value", f.message
      end
    end
  end

  def test_detects_typo
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "DATABASE_URL=x\n")
      File.write(File.join(dir, "app.rb"), "ENV[\"DATBASE_URL\"]\n")
      findings = Envdoctor::Scanner.scan(dir)
      typo = findings.find { |f| f.rule == "typo" }
      refute_nil typo
      assert_equal "DATBASE_URL", typo.name
      assert_equal "warning", typo.severity
      assert_equal "\"DATBASE_URL\" is not defined; did you mean \"DATABASE_URL\"?", typo.message
      # undefined-in-source still also fires for the same name.
      assert(findings.any? { |f| f.rule == "undefined-in-source" && f.name == "DATBASE_URL" })
    end
  end

  def test_detects_environment_diff
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "SHARED=1\nONLY_DEFAULT=1\n")
      File.write(File.join(dir, ".env.production"), "SHARED=1\nONLY_PROD=1\n")
      findings = Envdoctor::Scanner.scan(dir)
      diffs = findings.select { |f| f.rule == "environment-diff" }
      od = diffs.find { |f| f.name == "ONLY_DEFAULT" }
      refute_nil od
      assert_equal "warning", od.severity
      assert_equal "defined in default but missing in production", od.message
      assert(diffs.any? { |f| f.name == "ONLY_PROD" && f.message == "defined in production but missing in default" })
      # present in both → no diff.
      refute(diffs.any? { |f| f.name == "SHARED" })
    end
  end

  def test_detects_type_mismatch
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "PORT=8080\nWORKERS=4\n")
      File.write(File.join(dir, ".env.production"), "PORT=production\nWORKERS=8\n")
      findings = Envdoctor::Scanner.scan(dir)
      tm = findings.select { |f| f.rule == "type-mismatch" }
      names = tm.map(&:name)
      assert_includes names, "PORT" # integer vs string → error
      refute_includes names, "WORKERS" # two integers → no error
      port = tm.find { |f| f.name == "PORT" }
      assert_equal "error", port.severity
      assert_equal "inferred type differs across environments", port.message
    end
  end

  def test_json_shape_and_no_value_leak
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "API_KEY=changeme\nUNUSED_KEY=supersecretvalue\n")
      File.write(File.join(dir, "app.rb"), "ENV[\"API_KEY\"]\n")
      findings = Envdoctor::Scanner.scan(dir)
      json = Envdoctor::Scanner.to_json_array(findings)
      arr = JSON.parse(json)
      assert_kind_of Array, arr
      arr.each do |obj|
        assert_equal %w[file line message name rule severity], obj.keys.sort
      end
      # values must never leak into JSON output
      refute_includes json, "supersecretvalue"
      refute_includes json, "changeme"
    end
  end
  def test_scans_compose_actions_and_kubernetes
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "DB_URL=postgres://localhost/db\n")
      File.write(File.join(dir, "docker-compose.yml"), <<~YAML)
        services:
          web:
            environment:
              SECRET: ${COMPOSE_SECRET}
              DATABASE: ${DB_URL}
      YAML
      wf = File.join(dir, ".github", "workflows")
      FileUtils.mkdir_p(wf)
      File.write(File.join(wf, "ci.yml"), <<~YAML)
        jobs:
          deploy:
            steps:
              - run: deploy
                env:
                  KEY: ${{ secrets.DEPLOY_KEY }}
                  REGION: ${{ vars.REGION }}
      YAML
      File.write(File.join(dir, "deployment.yaml"), <<~YAML)
        apiVersion: apps/v1
        kind: Deployment
        spec:
          value: ${K8S_VAR}
      YAML

      findings = Envdoctor::Scanner.scan(dir)
      undef_found = findings.select { |f| f.rule == "undefined-in-source" }
      undef_names = undef_found.map(&:name)
      assert_includes undef_names, "COMPOSE_SECRET"
      assert_includes undef_names, "DEPLOY_KEY"
      assert_includes undef_names, "REGION"
      assert_includes undef_names, "K8S_VAR"
      undef_found.each do |f|
        assert_equal "referenced but not defined in any environment file", f.message
      end
      # DB_URL referenced by compose interpolation → not unused.
      refute(findings.any? { |f| f.rule == "unused" && f.name == "DB_URL" })
      # values never leak.
      json = Envdoctor::Scanner.to_json_array(findings)
      refute_includes json, "postgres://localhost/db"
    end
  end

  EXAMPLE = "# Generated by envdoctor. Fill in values; do not commit secrets.\nDB_URL=\nPORT=\n"

  def with_generate_project
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "DB_URL=secretvalue\n")
      File.write(File.join(dir, "app.rb"), "ENV[\"PORT\"]\n")
      yield dir
    end
  end

  def test_env_example_content_exact_and_no_values
    with_generate_project do |dir|
      content = Envdoctor::Scanner.env_example_content(dir)
      assert_equal EXAMPLE, content
      refute_includes content, "secretvalue"
    end
  end

  def test_environment_doc_content
    with_generate_project do |dir|
      content = Envdoctor::Scanner.environment_doc_content(dir)
      assert_includes content, "| DB_URL | yes | no |"
      assert_includes content, "| PORT | no | yes |"
      assert content.end_with?("\n")
      refute_includes content, "secretvalue"
    end
  end

  def test_init_creates_then_skips_then_force
    with_generate_project do |dir|
      out, = capture_io { Envdoctor::CLI.run(["init", "-d", dir]) }
      assert_includes out, "created .env.example"
      assert_includes out, "created ENVIRONMENT.md"
      assert_equal EXAMPLE, File.read(File.join(dir, ".env.example"))

      # second run without --force skips
      File.write(File.join(dir, ".env.example"), "TOUCHED\n")
      out2, = capture_io { Envdoctor::CLI.run(["init", "-d", dir]) }
      assert_includes out2, "skipped .env.example (exists)"
      assert_equal "TOUCHED\n", File.read(File.join(dir, ".env.example"))

      # --force rewrites
      out3, = capture_io { Envdoctor::CLI.run(["init", "-d", dir, "--force"]) }
      assert_includes out3, "wrote .env.example"
      assert_equal EXAMPLE, File.read(File.join(dir, ".env.example"))
    end
  end

  def test_fix_always_rewrites
    with_generate_project do |dir|
      File.write(File.join(dir, ".env.example"), "STALE\n")
      out, = capture_io { Envdoctor::CLI.run(["fix", "-d", dir]) }
      assert_includes out, "wrote .env.example"
      assert_includes out, "wrote ENVIRONMENT.md"
      assert_equal EXAMPLE, File.read(File.join(dir, ".env.example"))
      doc = File.read(File.join(dir, "ENVIRONMENT.md"))
      assert_includes doc, "| DB_URL | yes | no |"
      assert_includes doc, "| PORT | no | yes |"
    end
  end

  def test_diff_and_sync
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "A=1\nB=2\n")
      File.write(File.join(dir, ".env.production"), "A=9\n")
      d = Envdoctor::Scanner.diff_labels(dir, "default", "production")
      assert_equal({ "onlyInA" => ["B"], "onlyInB" => [], "common" => ["A"] }, d)
      assert_equal ["B"], Envdoctor::Scanner.sync_labels(dir, "default", "production", dry_run: true)
      refute_includes File.read(File.join(dir, ".env.production")), "B="
      assert_equal ["B"], Envdoctor::Scanner.sync_labels(dir, "default", "production")
      prod = File.read(File.join(dir, ".env.production"))
      assert_includes prod, "B=\n"
      assert_includes prod, "A=9"
      refute_includes prod, "B=2"
    end
  end

  def test_schema_validation
    Dir.mktmpdir do |dir|
      File.write(File.join(dir, ".env"), "PORT=99999\nLEVEL=verbose\nAPI=ftp://x\nGOOD=info\n")
      File.write(File.join(dir, "envdoctor.schema.json"),
                 %q({"PORT":{"type":"integer","max":65535},"LEVEL":{"enum":["debug","info"]},"API":{"type":"url"},"MISSING":{"type":"string"},"GOOD":{"enum":["info","warn"]}}))
      findings = Envdoctor::Scanner.scan(dir)
      schema = findings.select { |f| f.rule == "schema-validation" }.map { |f| [f.name, f.message] }.to_h
      assert_equal({ "PORT" => "value exceeds the maximum",
                     "LEVEL" => "value is not one of the allowed values",
                     "API" => "value does not match schema type url",
                     "MISSING" => "required by schema but not defined" }, schema)
      refute schema.key?("GOOD")
    end
  end

end

File: python/tests/test_scanner.py
import json
from pathlib import Path

from envdoctor.cli import main
from envdoctor.scanner import (
    finding_to_dict,
    infer_type,
    levenshtein,
    scan,
    scan_source_file,
)


def _write(tmp_path: Path, name: str, content: str) -> Path:
    p = tmp_path / name
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(content)
    return p


def test_detects_all_usage_forms(tmp_path):
    src = _write(
        tmp_path,
        "app.py",
        "import os\n"
        "from os import environ\n"
        "# os.getenv('COMMENTED')\n"
        "a = os.getenv('DB_URL')\n"
        "b = os.environ.get('PORT')\n"
        "c = os.environ['API_KEY']\n"
        "d = environ.get('HOST')\n"
        "e = environ['DB_USER']\n"
        '"""docstring os.getenv("DOC_IGNORED")"""\n',
    )
    used = scan_source_file(src)
    assert set(used) == {"DB_URL", "PORT", "API_KEY", "HOST", "DB_USER"}
    assert "COMMENTED" not in used
    assert "DOC_IGNORED" not in used


def test_missing_and_unused(tmp_path):
    _write(tmp_path, ".env", "DB_URL=postgres://x\nUNUSED_KEY=1\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('DB_URL')\nos.getenv('NEW_FLAG')\n")

    result = scan(tmp_path)
    errors = {f.name for f in result.errors}
    warnings = {f.name for f in result.warnings}

    assert "NEW_FLAG" in errors  # used but not defined
    assert "UNUSED_KEY" in warnings  # defined but not used
    assert "DB_URL" not in errors and "DB_URL" not in warnings  # reconciled


def test_duplicates(tmp_path):
    _write(tmp_path, ".env", "DB_URL=a\nDB_URL=b\nSOLO=1\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('DB_URL')\nos.getenv('SOLO')\n")

    result = scan(tmp_path)
    dups = [f for f in result.findings if f.rule == "duplicates"]
    assert len(dups) == 1
    assert dups[0].name == "DB_URL"
    assert dups[0].severity == "error"
    assert "lines 1, 2" in dups[0].message
    # Single-definition key is not reported as a duplicate.
    assert "SOLO" not in {f.name for f in dups}
    # First occurrence still reconciles: DB_URL is not undefined/unused.
    assert "DB_URL" not in {f.name for f in result.warnings}


def test_public_prefix(tmp_path):
    _write(
        tmp_path,
        ".env",
        "NEXT_PUBLIC_API_KEY=x\nPUBLIC_URL=x\nAPI_KEY=x\n",
    )
    result = scan(tmp_path)
    pp = [f for f in result.findings if f.rule == "public-prefix"]
    names = {f.name for f in pp}
    assert names == {"NEXT_PUBLIC_API_KEY"}
    assert pp[0].severity == "error"
    assert "PUBLIC_URL" not in names  # public prefix but not secret-like
    assert "API_KEY" not in names  # secret-like but no public prefix


def test_clean_project_has_no_findings(tmp_path):
    _write(tmp_path, ".env", "DB_URL=x\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('DB_URL')\n")
    result = scan(tmp_path)
    assert result.findings == []


def test_infer_type():
    assert infer_type("") == "empty"
    assert infer_type("3000") == "integer"
    assert infer_type("-7") == "integer"
    assert infer_type("1.5") == "float"
    assert infer_type("TRUE") == "boolean"
    assert infer_type("https://x.io") == "url"
    assert infer_type('{"a":1}') == "json"
    assert infer_type("hello") == "string"


def test_levenshtein():
    assert levenshtein("abc", "abc") == 0
    assert levenshtein("DATBASE_URL", "DATABASE_URL") == 1


def test_weak_secret(tmp_path):
    _write(
        tmp_path,
        ".env",
        "API_KEY=changeme\nSTRONG_TOKEN=s0m3-l0ng-r4nd0m-value\nSHORT_SECRET=abc\n",
    )
    _write(tmp_path, "app.py", "import os\n")
    result = scan(tmp_path)
    weak = {f.name for f in result.findings if f.rule == "weak-secret"}
    assert "API_KEY" in weak  # placeholder
    assert "SHORT_SECRET" in weak  # too short
    assert "STRONG_TOKEN" not in weak
    for f in result.findings:
        assert f.severity in ("error", "warning")
        # value must never appear anywhere in a finding
        assert "changeme" not in f.message
        assert "s0m3" not in f.message


def test_typo_suggestion(tmp_path):
    _write(tmp_path, ".env", "DATABASE_URL=postgres://x\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('DATBASE_URL')\n")
    result = scan(tmp_path)
    typos = [f for f in result.findings if f.rule == "typo"]
    assert len(typos) == 1
    assert typos[0].name == "DATBASE_URL"
    assert typos[0].severity == "warning"
    assert 'did you mean "DATABASE_URL"' in typos[0].message
    # undefined-in-source still fires too.
    assert "DATBASE_URL" in {f.name for f in result.errors}


def test_environment_diff(tmp_path):
    _write(tmp_path, ".env", "SHARED=1\n")
    _write(tmp_path, ".env.production", "SHARED=1\nONLY_PROD=1\n")
    _write(tmp_path, "app.py", "import os\n")
    result = scan(tmp_path)
    diffs = {f.name: f for f in result.findings if f.rule == "environment-diff"}
    assert "ONLY_PROD" in diffs
    assert diffs["ONLY_PROD"].severity == "warning"
    assert "default" in diffs["ONLY_PROD"].message
    assert "production" in diffs["ONLY_PROD"].message
    assert "SHARED" not in diffs  # present in both


def test_type_mismatch(tmp_path):
    _write(tmp_path, ".env", "PORT=3000\nHOST=8080\n")
    _write(tmp_path, ".env.production", "PORT=abc\nHOST=9090\n")
    _write(tmp_path, "app.py", "import os\n")
    result = scan(tmp_path)
    mismatches = {f.name for f in result.findings if f.rule == "type-mismatch"}
    assert "PORT" in mismatches  # integer vs string
    assert "HOST" not in mismatches  # integer vs integer, same group
    tm = next(f for f in result.findings if f.rule == "type-mismatch")
    assert tm.severity == "error"
    assert "3000" not in tm.message and "abc" not in tm.message


def test_json_output_shape(tmp_path, capsys):
    _write(tmp_path, ".env", "API_KEY=changeme\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('DATBASE_URL')\n")
    _write(tmp_path, ".env.production", "API_KEY=changeme\nDATABASE_URL=x\n")
    code = main(["scan", "--dir", str(tmp_path), "--json"])
    assert code == 1  # undefined-in-source is an error
    out = capsys.readouterr().out
    data = json.loads(out)
    assert isinstance(data, list)
    for obj in data:
        assert set(obj) == {"rule", "severity", "name", "message", "file", "line"}
    # no value strings leak
    assert "changeme" not in out


def test_finding_to_dict_shape(tmp_path):
    _write(tmp_path, ".env", "UNUSED=1\n")
    _write(tmp_path, "app.py", "import os\n")
    result = scan(tmp_path)
    d = finding_to_dict(result.findings[0], tmp_path)
    assert d["file"] == ".env"
    assert isinstance(d["line"], int)


def test_infra_sources_scanned(tmp_path):
    _write(tmp_path, ".env", "DB_URL=postgres://x\n")
    _write(
        tmp_path,
        "docker-compose.yml",
        "services:\n"
        "  web:\n"
        "    image: x\n"
        "    environment:\n"
        "      - TOKEN=${COMPOSE_SECRET}\n"
        "      - DB=${DB_URL}\n",
    )
    _write(
        tmp_path,
        ".github/workflows/ci.yml",
        "jobs:\n"
        "  build:\n"
        "    steps:\n"
        "      - run: deploy\n"
        "        env:\n"
        "          KEY: ${{ secrets.DEPLOY_KEY }}\n"
        "          REGION: ${{ vars.REGION }}\n",
    )
    _write(
        tmp_path,
        "k8s/deploy.yaml",
        "apiVersion: apps/v1\n"
        "kind: Deployment\n"
        "spec:\n"
        "  value: ${K8S_VAR}\n",
    )

    result = scan(tmp_path)
    undefined = {
        f.name
        for f in result.errors
        if f.rule == "undefined-in-source"
    }
    # Referenced only in infra files but never defined -> undefined-in-source.
    assert {"COMPOSE_SECRET", "DEPLOY_KEY", "REGION", "K8S_VAR"} <= undefined
    # New message wording.
    for f in result.findings:
        if f.rule == "undefined-in-source":
            assert f.message == "referenced but not defined in any environment file"
    # DB_URL is defined and referenced in compose -> NOT reported unused.
    unused = {f.name for f in result.warnings if f.rule == "unused"}
    assert "DB_URL" not in unused
    # No values leak into any finding.
    for f in result.findings:
        assert "postgres" not in f.message


def test_infra_no_value_leak(tmp_path):
    _write(tmp_path, ".env", "SECRET_TOKEN=s0m3-l0ng-r4nd0m-value\n")
    _write(
        tmp_path,
        "compose.yaml",
        "services:\n  a:\n    environment:\n      X: ${SECRET_TOKEN}\n",
    )
    result = scan(tmp_path)
    for f in result.findings:
        assert "s0m3" not in f.message


def test_generate_docs_exact_content(tmp_path):
    from envdoctor.scanner import generate_docs

    _write(tmp_path, ".env", "DB_URL=postgres://x\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('PORT')\n")

    docs = generate_docs(tmp_path)
    assert docs[".env.example"] == (
        "# Generated by envdoctor. Fill in values; do not commit secrets.\n"
        "DB_URL=\n"
        "PORT=\n"
    )
    md = docs["ENVIRONMENT.md"]
    assert md.startswith("# Environment variables\n\n| Variable | Defined | Used |\n")
    assert "| DB_URL | yes | no |" in md
    assert "| PORT | no | yes |" in md
    assert md.endswith("\n")
    # Values must never leak.
    assert "postgres" not in docs[".env.example"]
    assert "postgres" not in md


def test_init_creates_skips_and_forces(tmp_path, capsys):
    _write(tmp_path, ".env", "DB_URL=postgres://x\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('PORT')\n")

    example = tmp_path / ".env.example"
    envdoc = tmp_path / "ENVIRONMENT.md"

    # First init creates both.
    assert main(["init", "--dir", str(tmp_path)]) == 0
    out = capsys.readouterr().out
    assert "created .env.example" in out
    assert "created ENVIRONMENT.md" in out
    first_example = example.read_text()
    first_md = envdoc.read_text()
    assert first_example == (
        "# Generated by envdoctor. Fill in values; do not commit secrets.\n"
        "DB_URL=\nPORT=\n"
    )
    assert "postgres" not in first_example and "postgres" not in first_md

    # Second init without --force skips; files unchanged.
    example.write_text("SENTINEL")
    assert main(["init", "--dir", str(tmp_path)]) == 0
    out = capsys.readouterr().out
    assert "skipped .env.example (exists)" in out
    assert "skipped ENVIRONMENT.md (exists)" in out
    assert example.read_text() == "SENTINEL"

    # --force rewrites.
    assert main(["init", "--dir", str(tmp_path), "--force"]) == 0
    out = capsys.readouterr().out
    assert "wrote .env.example" in out
    assert "wrote ENVIRONMENT.md" in out
    assert example.read_text() == first_example


def test_fix_always_rewrites(tmp_path, capsys):
    _write(tmp_path, ".env", "DB_URL=postgres://x\n")
    _write(tmp_path, "app.py", "import os\nos.getenv('PORT')\n")

    example = tmp_path / ".env.example"
    example.write_text("STALE")
    assert main(["fix", "--dir", str(tmp_path)]) == 0
    out = capsys.readouterr().out
    assert "wrote .env.example" in out
    assert "wrote ENVIRONMENT.md" in out
    assert example.read_text() == (
        "# Generated by envdoctor. Fill in values; do not commit secrets.\n"
        "DB_URL=\nPORT=\n"
    )
    md = (tmp_path / "ENVIRONMENT.md").read_text()
    assert "| DB_URL | yes | no |" in md
    assert "| PORT | no | yes |" in md
    assert "postgres" not in example.read_text() and "postgres" not in md


def test_diff_and_sync(tmp_path):
    from envdoctor.scanner import diff_labels, sync_labels

    (tmp_path / ".env").write_text("A=1\nB=2\n")
    (tmp_path / ".env.production").write_text("A=9\n")

    d = diff_labels(tmp_path, "default", "production")
    assert d == {"onlyInA": ["B"], "onlyInB": [], "common": ["A"]}

    # dry-run does not write
    assert sync_labels(tmp_path, "default", "production", dry_run=True) == ["B"]
    assert "B=" not in (tmp_path / ".env.production").read_text()

    # real sync appends B= (no value copied) and leaves A's value intact
    assert sync_labels(tmp_path, "default", "production") == ["B"]
    prod = (tmp_path / ".env.production").read_text()
    assert "B=\n" in prod and "A=9" in prod and "B=2" not in prod
    assert diff_labels(tmp_path, "default", "production")["common"] == ["A", "B"]


def test_schema_validation(tmp_path):
    (tmp_path / ".env").write_text("PORT=99999\nLEVEL=verbose\nAPI=ftp://x\nGOOD=info\n")
    (tmp_path / "envdoctor.schema.json").write_text(
        '{"PORT":{"type":"integer","max":65535},"LEVEL":{"enum":["debug","info"]},'
        '"API":{"type":"url"},"MISSING":{"type":"string"},"GOOD":{"enum":["info","warn"]}}'
    )
    result = scan(tmp_path)
    schema = {f.name: f.message for f in result.findings if f.rule == "schema-validation"}
    assert schema == {
        "PORT": "value exceeds the maximum",
        "LEVEL": "value is not one of the allowed values",
        "API": "value does not match schema type url",
        "MISSING": "required by schema but not defined",
    }
    assert "GOOD" not in schema  # valid, no finding
    # no values leak
    for f in result.findings:
        assert "99999" not in f.message and "verbose" not in f.message


def test_schema_optional_and_absent(tmp_path):
    from envdoctor.scanner import load_schema

    (tmp_path / ".env").write_text("A=1\n")
    # no schema file → no schema findings
    assert load_schema(tmp_path) == {}
    assert not [f for f in scan(tmp_path).findings if f.rule == "schema-validation"]
    # optional missing var → suppressed
    (tmp_path / "envdoctor.schema.json").write_text('{"MAYBE":{"type":"string","optional":true}}')
    assert not [f for f in scan(tmp_path).findings if f.rule == "schema-validation"]
